### PR TITLE
Extract BaseEvent class

### PR DESCRIPTION
**Pull request by athill**
* Extract most of DropdownEvent into BaseEvent
* Move `keys` into BaseEvent

### CODE REVIEW

**Comment by jhoerr** [comment]
Primo refactor, @athill

File: src/components/util/BaseEvent.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
+
+
+export default abstract class BaseEvent {
```
Comment (jwbennet) [inline src/components/util/BaseEvent.ts]: How about renaming this to something more descriptive?  Maybe `AbstractUserActionEvent`?

File: src/components/Dropdown/Dropdown.tsx [hunk 2]
```diff
@@ -36,7 +36,7 @@ export class Dropdown extends React.PureComponent<DropdownProps & React.HTMLAttr
         this.toggleButton = React.createRef();
         this.handleClickOutside = this.handleClickOutside.bind(this);
         this.toggleDropdown = this.toggleDropdown.bind(this);
-        this.eventHandler = DropdownEvent.handler(this.handleClickOutside);
+        this.eventHandler = DropdownEvent.handler(this.handleClickOutside, DropdownEvent);
```
Comment (jwbennet) [inline src/components/Dropdown/Dropdown.tsx]: How about we provide another version of this function in `DropdownEvent` which provides the second argument so the code that uses it does not have to?

**Comment by athill** [comment]
@jwbennet Requested changes made

File: src/components/util/AbstractUserActionEvent.test.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+const createMouseEvent = (buttonCode) => ({
+    type: 'click',
+    which: buttonCode
+}) as MouseEvent;
+
+const createTouchEvent = () => ({
+    type: 'touchstart'
+}) as TouchEvent;
+
+describe('DropdownEvent', () => {
```
Comment (jwbennet) [inline src/components/util/AbstractUserActionEvent.test.ts]: This probably needs to be changed to `AbstractUserActionEvent`

**Pull request merged**